### PR TITLE
Improve DxCore support

### PR DESCRIPTION
**Pull request by brunob45**
- Added support for `pio run -t bootloader` command
- Added an option to set the `MILLIS_USE_TIMER` variable: `board_hardware.millistimer` (see description in SpenceKonde/DxCore#333)
- Added the build flag `USING_OPTIBOOT` to all DxCore variants when a bootloader is specified (following SpenceKonde/DxCore#329)
- ~Added board definition file for future AVR-DD & EA chips~

This PR may help closing issues #41 & #42.

### CODE REVIEW

**Comment by matou78** [comment]
oh nice work :D

**Comment by matou78** [comment]
hello @brunob45 ,
Found finally some time to test your things but it didn't change 
Still the error message when trying to burn the bootloader to a AVR64DA32 or AVR64DA64 chips ( the command 'pio run -t bootloader' is working tough )

The flag 'USING_OPTIBOOT' not working, i still need to add it to the 'build_flag=' myself to platformio.ini if i want to use bootloader

Maybe it's not finished ? i don't know much about this kind of code

**Comment by brunob45** [comment]
Hello @matou78 ,

May I get a copy of your `platformio.ini` config file to try it out?

**Comment by matou78** [comment]
Sorry for the delay, too much work ...
```
[env]
platform = atmelmegaavr
board = AVR64DA32
framework = arduino
board_build.f_cpu = 24000000L

[env:Upload_UART]
;build_flags = -DUSING_OPTIBOOT
upload_protocol = arduino
upload_speed = 115200

[env:serial_updi]
upload_protocol = serialupdi
upload_flags =  -e 
                -v
```

**Comment by brunob45** [comment]
I just pushed a commit to use the latest version of avrdude.
Could you try this config?

```
[env]
; platform is using this PR
platform = https://github.com/brunob45/platform-atmelavrdx#dev/avrea
board = AVR64DA32
framework = arduino
board_build.f_cpu = 24000000L

; Set bootloader serial port : ser0, ser1, or ser2
; For uploads with manual reset (no autoreset circuit), add <_8sec> to the selected port (ex. ser1_8sec)
board_hardware.uart = ser1_8sec

[env:Upload_UART]
upload_protocol = arduino
upload_speed = 115200

[env:serial_updi]
upload_protocol = serialupdi
upload_flags = -e -v
```

Then, to load the bootloader
`pio run -e serial_updi -t bootloader`
and to upload the code
`pio run -e Upload_UART -t upload`

By using the `_8sec` bootloader, I'm able to power cycle the chip, then quickly launch the `upload` target. It will execute within this 8 seconds timeout and upload successfully. Be sure to select the correct USART port (0,1, or 2) in the platformio.ini file.

**Comment by matou78** [comment]
Thanks :) will try it asap

**Comment by matou78** [comment]
@brunob45 Just gave it a try , there is few wrong things i found

the name of the .hex files that the builder is calling is wrong , for example the file called is `optiboot_dx64_ser1_8sec.hex` but the correct one found in the framework is `optiboot_64dx_ser1_all_8sec.hex`

after i renamed the .hex so it get called without error ,when setting fuses , avrdude show error 
```
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/atmelmegaavr/AVR64DA32.html
PLATFORM: Atmel megaAVR (1.7.0+sha.e5ca771) > AVR64DA32
HARDWARE: AVR64DA32 24MHz, 8KB RAM, 64KB Flash
PACKAGES:
 - framework-arduino-megaavr-dxcore @ 1.4.10
 - tool-avrdude @ 1.70100.0 (7.1.0)
 - toolchain-atmelavr @ 3.70300.220127 (7.3.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 18 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Using bootloader optiboot_dx64_ser1_8sec.hex

TARGET CONFIGURATION:
-------------------------
Target = avr64da32
Clock speed = 24000000L
Oscillator = internal
BOD level = 2.6v
Save EEPROM = yes
Reset pin mode = reset
-------------------------
Auto-detected: COM7

Selected fuses:
-------------------------
[fuse0 / wdtcfg   = 0x00]
[fuse1 / bodcfg   = 0x00]
[fuse2 / osccfg   = 0x00]
[fuse4 / tcd0cfg  = 0x00]
[fuse5 / syscfg0  = 0xC9]
[fuse6 / syscfg1  = 0x06]
[fuse7 / codesize = 0x00]
[fuse8 / bootsize = 0x01]
[lock  / lockbit  = 0x5CC5C55C]
-------------------------

Setting fuses...
avrdude: unknown option --  
```

Thanks for your time

**Comment by brunob45** [comment]
Please check the Platformio documentation in the [official DxCore repo](https://github.com/SpenceKonde/DxCore/). There's a section about [how to select the bootloader](https://github.com/SpenceKonde/DxCore/blob/master/PlatformIO.md#board_hardwareuart).

## Comparison with Arduino IDE
The `board_hardware.uart = ...` is where you tell if you are using Optiboot or not.

- If you are uploading via UPDI, then you are <ins>not</ins> using Optiboot, so leave this line commented out.
  This is the same as selecting the board `AVR DA-series (no bootloader)` in the Arduino IDE `Tools`->`Board: ...`
  Note that in the Arduino IDE, the menus `Tools`->`Bootloader Serial Port` and `Bootloader Entry Condition` are not present when this board is selected.

- If you are uploading via serial (arduino), then tell which Serial Port you'll be using.
  This is the same as selecting the board `AVR DA-series (Optiboot)` in the Arduino IDE `Tools`->`Board: ...`
  In the Arduino IDE, you would normally select the bootloader in the menu `Tools`->`Bootloader Serial Port` and `Bootloader Entry Condition`.
  In PlatformIO, you specify the bootloader using `board_hardware.uart = ...` (ex. `ser1_8sec`).

Note that, in the Arduino IDE, the `upload_port` is dependant on the selected board: `Optiboot` or `no bootloader`.
- If `Optiboot`, the `upload_port` is the equivalent of PlatformIO's `arduino`.
- If `no bootloader`, the `upload_port` is the protocol chosen in `Tools`->`Programmer: ...`.

## In other words
If you want to upload via serial
- Set `board_hardware.uart = ...`.
- Burn the bootloader.
- You will not be able to upload via UPDI anymore.

if you want to upload via UPDI
- Comment out `board_hardware.uart = ...`.
- Reset the fuses `pio run -t fuses`.
- You will not be able to upload via serial anymore.

**Comment by valeros** [comment]
Hi @brunob45, thanks for the PR. It seems the latest v1.5.6 follows a different bootloader file name pattern, e.g. `optiboot_32dd14_ser0_alt3_poronly_8sec.hex`. Any thoughts?

**Comment by matou78** [comment]
oh shit this is why i didn't understand why i had to rename myself the files x)
i updated the dxcore to the latest version manually

**Comment by brunob45** [comment]
@valeros
I noticed that too.
I thought about different solutions. Should it be backward compatible?

If the dxcore package is updated to 1.5.x, we could just update the bootloader file name pattern. I already fixed it in my branch `dev/avrea2` https://github.com/brunob45/platform-atmelavrdx/tree/dev/avrea2

**Comment by brunob45** [comment]
@valeros
I noticed that too.
I thought about different solutions. Should it be backward compatible?

If the dxcore package is updated to 1.5.x, we could just update the bootloader file name pattern. I already fixed it in my branch `dev/avrea2` https://github.com/brunob45/platform-atmelavrdx/tree/dev/avrea2

But this won't work for the DD series... Let me try something first

**Comment by brunob45** [comment]
@valeros 

I edited the `bootloader.py` script to be able to specify the bootloader. For example, in `platformio.ini`:
```ini
[env]
...
board_bootloader.class = optiboot_32dd14
board_bootloader.port = ser0_alt3         ; Serial Port used for upload
board_bootloader.entrycond = poronly_8sec ; Entry Condition to enable upload
...
```

The resulting bootloader file would be `optiboot_32dd14_ser0_alt3_poronly_8sec.hex`.

In another PR, I would like to edit the `.json` files for all DxCore boards, adding
```json
{
  ...
  "bootloader": {
    "class": "optiboot_32dx"
  },
  ...
}
```
so the default value for `bootloader.class` is correct.
See:
- For DxCore 1.4.x: https://github.com/brunob45/platform-atmelavrdx/compare/dev/avrea...brunob45:platform-atmelavrdx:dev/dxcore-btld-1.4
- For DxCore 1.5.x: https://github.com/brunob45/platform-atmelavrdx/compare/dev/avrea...brunob45:platform-atmelavrdx:dev/dxcore-btld

What do you think about this solution?

**Review by valeros — CHANGES_REQUESTED** [review]
Please see my comments.

File: builder/bootloader.py [hunk 1]
```diff
@@ -44,6 +44,24 @@ def get_suitable_optiboot_binary(framework_dir, board_config):
     return bootloader_path
 
 
+def get_bootloader_dxcore(framework_dir, board_config):
+    # print(json.dumps(board_config.manifest, indent=2)) # print board config content
```
[the remark this answers is not in the document]
Comment (brunob45) [reply builder/bootloader.py]: Line removed

File: builder/bootloader.py [hunk 2]
```diff
@@ -44,6 +44,24 @@ def get_suitable_optiboot_binary(framework_dir, board_config):
...
+    port = board_config.get("bootloader.port", "no_bootloader").lower()
+    if port == "no_bootloader":
+        return ""
+
+    btld = board_config.get("bootloader.class")
+    entry = board_config.get("bootloader.entrycond", "extr") # use "extr" by default
+    bootloader_file = f"{btld}_{port}_{entry}.hex"
```
[the remark this answers is not in the document]
Comment (brunob45) [reply builder/bootloader.py]: I added a check for valid `bootloader.class` and `bootloader.port`

**Comment by brunob45** [comment]
@valeros 
Thank you for the comments, the issues have been fixed

File: builder/bootloader.py [hunk 3]
```diff
@@ -44,6 +44,35 @@ def get_suitable_optiboot_binary(framework_dir, board_config):
...
+
+    if not btld:
+        sys.stderr.write("Error: invalid `bootloader.class` in board config!\n")
+        env.Exit(1)
+    if not port:
+        sys.stderr.write("Error: invalid `bootloader.port` in board config!\n")
+        env.Exit(1)
+
+    if entry:
+        bootloader_file = f"{btld}_{port}_{entry}.hex"
```
Comment (valeros) [inline builder/bootloader.py:60]: Are you sure this pattern will properly work with the bootloader binaries from the latest 1.5.6? For example `optiboot_32dd14_ser0_alt3_extr_8sec.hex`.

**Review by valeros — CHANGES_REQUESTED** [review]

File: builder/bootloader.py (hunk 3 above)
Comment (brunob45) [reply builder/bootloader.py:60]: I confirm that the example `optiboot_32dd14_ser0_alt3_extr_8sec.hex` is working with the latest version of DxCore.
```ini
[env]
board_bootloader.class = optiboot_32dd14
board_bootloader.port = ser0_alt3
board_bootloader.entrycond = extr_8sec
```

However, the `toolchain-atmelavr@3.70300.220127` does not support the AVR16DDxx and AVR32DDxx. Only AVR64DDxx are supported.
